cpp/include/cudf/ast/linearizer.hpp

```diff
@@ -0,0 +1,233 @@
+ *
+ * This determines which table to use in cases with two tables (e.g. joins).
+ *
+ */
+enum class table_reference {
+  LEFT,   // Column index in the left table
+  RIGHT,  // Column index in the right table
+  OUTPUT  // Column index in the output table
+};
+
+// Forward declaration
+class literal;
+class column_reference;
+class expression;
+
+/**
+ * @brief A literal value used in an abstract syntax tree.
+ *
+ */
+class literal : public detail::node {
+  friend class detail::linearizer;
+
+ public:
+  /**
+   * @brief Construct a new literal object.
+   *
+   * @param value A fixed width scalar device view containing the literal value.
+   */
+  literal(const cudf::detail::fixed_width_scalar_device_view_base value) : value(value) {}
+
+  /**
+   * @brief Get the data type.
+   *
+   * @return cudf::data_type
+   */
+  cudf::data_type get_data_type() const { return this->get_value().type(); }
+
+ private:
+  /**
+   * @brief Get the value object.
+   *
+   * @return cudf::detail::fixed_width_scalar_device_view_base
+   */
+  cudf::detail::fixed_width_scalar_device_view_base get_value() const { return this->value; }
+
+  /**
+   * @brief Accepts a visitor class.
+   *
+   * @param visitor Visitor.
+   * @return cudf::size_type Index of device data reference for this instance.
+   */
+  cudf::size_type accept(detail::linearizer& visitor) const override;
+
+  const cudf::detail::fixed_width_scalar_device_view_base value;
+};
+
+/**
+ * @brief A node referring to data from a column in a table.
+ *
+ */
+class column_reference : public detail::node {
+  friend class detail::linearizer;
+
+ public:
+  /**
+   * @brief Construct a new column reference object
+   *
+   * @param column_index Index of this column in the table (provided when the node is
+   * evaluated).
+   * @param table_source Which table to use in cases with two tables (e.g. joins).
+   */
+  column_reference(cudf::size_type column_index,
+                   table_reference table_source = table_reference::LEFT)
+    : column_index(column_index), table_source(table_source)
+  {
+  }
+
+  /**
+   * @brief Get the column index.
+   *
+   * @return cudf::size_type
+   */
+  cudf::size_type get_column_index() const { return this->column_index; }
+
+  /**
+   * @brief Get the table source.
+   *
+   * @return table_reference
+   */
+  table_reference get_table_source() const { return this->table_source; }
+
+  /**
+   * @brief Get the data type.
+   *
+   * @param table Table used to determine types.
+   * @return cudf::data_type
+   */
+  cudf::data_type get_data_type(const table_view& table) const
+  {
+    return table.column(this->get_column_index()).type();
+  }
+
+  /**
+   * @brief Get the data type.
+   *
+   * @param left_table Left table used to determine types.
+   * @param right_table Right table used to determine types.
+   * @return cudf::data_type
+   */
+  cudf::data_type get_data_type(const table_view& left_table, const table_view& right_table) const
```

Comment:
@jrhemstad Some of this code was originally written to support two input tables (for the case of inequality joins). Should I remove code to support two tables in this PR? (such as the `enum class table_reference` and device data references that pertain to left/right/output tables)

Reply:
I'm fine leaving some future looking code, especially since it's pretty small.